### PR TITLE
WLP improvements (#650) [6455]

### PR DESCRIPTION
**Pull request by LuisGP**
Backport of #650

### CODE REVIEW

File: src/cpp/qos/ReaderQos.cpp [hunk 1]
```diff
@@ -26,94 +26,89 @@ namespace fastrtps {
...
+    if(first_time)
+    {
+        m_reliability = qos.m_reliability;
+        m_reliability.hasChanged = true;
+    }
+    if(first_time)
+    {
+        m_ownership = qos.m_ownership;
+        m_ownership.hasChanged = true;
+    }
```
Comment (raquelalvarezbanos) [inline src/cpp/qos/ReaderQos.cpp]: The above three `if`s could be combined, but I assume as this is a backport we shouldn't aim at making any improvements?

File: test/blackbox/BlackboxTestsLivelinessQos.cpp [hunk 2]
```diff
@@ -32,7 +32,7 @@ TEST(LivelinessQos, Liveliness_Automatic_Reliable)
     // Liveliness lease duration and announcement period
     uint32_t liveliness_ms = 200;
     Duration_t liveliness_s(liveliness_ms * 1e-3);
-    Duration_t announcement_period(liveliness_ms * 1e-3 * 0.5);
+    Duration_t announcement_period(liveliness_ms * 1e-3 * 0.01);
```
Comment (raquelalvarezbanos) [inline test/blackbox/BlackboxTestsLivelinessQos.cpp]: As this is just a backport, I guess it is fine to leave these timings, although they are not enough to make tests stable.

File: src/cpp/rtps/builtin/discovery/participant/PDPSimple.cpp [hunk 3]
```diff
@@ -256,14 +256,19 @@ bool PDPSimple::initPDP(RTPSParticipantImpl* part)
...
 }
 
+bool PDPSimple::enable()
+{
+    bool result = mp_RTPSParticipant->enableReader(mp_SPDPReader);
+    if (result)
+    {
+        mp_resendParticipantTimer = new ResendParticipantProxyDataPeriod(this, m_discovery);
+    }
+    return result;
```
Comment (raquelalvarezbanos) [inline src/cpp/rtps/builtin/discovery/participant/PDPSimple.cpp]: Master is different now, we are not resending the participant proxy data, can you check when and why this was removed? I couldn't find it...

**Comment by raquelalvarezbanos** [comment]
A lot of liveliness tests failed on Windows. Maybe we should also backport [this](https://github.com/eProsima/Fast-RTPS/pull/689) to ensure that this backport is OK?

File: test/blackbox/BlackboxTestsLivelinessQos.cpp (hunk 2 above)
Comment (LuisGP) [reply test/blackbox/BlackboxTestsLivelinessQos.cpp]: Outdated dut to #689 backport.

File: src/cpp/qos/ReaderQos.cpp (hunk 1 above)
Comment (LuisGP) [reply src/cpp/qos/ReaderQos.cpp]: I doubt about this kind of things. I saw other parts of the code that may be improved, but as you said, as this is a backport I don't think we should waste time in this kind of things.

Comment (raquelalvarezbanos) [reply src/cpp/qos/ReaderQos.cpp]: Yeap, I agree, we should leave it as it is.

File: src/cpp/rtps/builtin/discovery/participant/PDPSimple.cpp (hunk 3 above)
Comment (LuisGP) [reply src/cpp/rtps/builtin/discovery/participant/PDPSimple.cpp]: This is needed in the backport due to the lack of the event system. In the new code was moved to `PDP::initPDP`.

**Comment by raquelalvarezbanos** [comment]
OK, so we've seen some tests time out and I think I found the reason: `WLivelinessPeriodicAssertion` needs to be updated as it is populating the liveliness message wrong. In the constructor, you just have to copy what we do in `WLP::send_liveliness_message()`. The problem was that `WLP` was fixed to serialize the payload correctly, but `WLivelinessPeriodicAssertion` wasn't.

File: src/cpp/rtps/builtin/discovery/participant/PDPSimple.cpp (hunk 3 above)
Comment (raquelalvarezbanos) [reply src/cpp/rtps/builtin/discovery/participant/PDPSimple.cpp]: :+1:

**Comment by LuisGP** [comment]
> OK, so we've seen some tests time out and I think I found the reason: `WLivelinessPeriodicAssertion` needs to be updated as it is populating the liveliness message wrong. In `WLivelinessPeriodicAssertion::add_cache_change()`, you just have to copy what we do in `WLP::send_liveliness_message()`. The problem was that `WLP` was fixed to serialize the payload correctly, but `WLivelinessPeriodicAssertion` wasn't.

Looks pretty good now. Thank you!

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/509/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/455/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/505/)

**Comment by rotu** [comment]
Just gave this branch a try at the suggestion of @JaimeMartin.
Totally solves [the bringup hitches I observed](https://blog.roverrobotics.com/navigation2-now-were-getting-somewhere/)! Good job, guys!

**Pull request merged**